beacon/goclient/goclient.go

```diff
@@ -141,19 +142,40 @@ var _ NodeClientProvider = (*GoClient)(nil)
 
 // GoClient implementing Beacon struct
 type GoClient struct {
-	log                  *zap.Logger
-	ctx                  context.Context
-	network              beaconprotocol.Network
-	client               Client
-	nodeVersion          string
-	nodeClient           NodeClient
-	gasLimit             uint64
-	operatorDataStore    operatordatastore.OperatorDataStore
+	log         *zap.Logger
+	ctx         context.Context
+	network     beaconprotocol.Network
+	client      Client
+	nodeVersion string
+	nodeClient  NodeClient
+	gasLimit    uint64
+
+	operatorDataStore operatordatastore.OperatorDataStore
+
 	registrationMu       sync.Mutex
 	registrationLastSlot phase0.Slot
 	registrationCache    map[phase0.BLSPubKey]*api.VersionedSignedValidatorRegistration
-	commonTimeout        time.Duration
-	longTimeout          time.Duration
+
+	// attestationReqMuPool helps us prevent the sending of multiple attestation data requests
+	// for the same slot number (to avoid doing unnecessary work).
+	// It's a pool of mutexes (not a single mutex) to allow for some parallelism when requests
+	// targeting different slots are made.
+	attestationReqMuPool [32]sync.Mutex
+	// attestationDataCache stores attestation data from Beacon node for a bunch of recently made
+	// requests (by slot number). It allows for requesting attestation data once per slot from
+	// Beacon node as well as always having/observing the same consistent data in any given slot
+	// (eg, Beacon node might return different merkle root for the same slot for different requests,
+	// and this can lead to undesirable effects for GoClient users).
+	// attestationDataCache is used concurrently, hence thread-safe map.
+	// Note, we cache attestation data by slot (and not slot+committee_index) because it's the same
+	// data across all 64 Ethereum committees assigned for each slot.
+	attestationDataCache *hashmap.Map[phase0.Slot, *phase0.AttestationData]
```

Comment:
We usually use the `ttlcache` package for caches that require pruning old values, perhaps we could consider using it here as well to avoid having `pruneStaleAttestationDataRunner`

Reply:
Done, unit-tests don't complain about it :)

Reply:
Yeah, I agree, let's make them uniform, and if we start seeing `ttlcache` on pprof, then we'll consider optimizing them all at once. I think almost all cases of its usage are similar: Cache something for some time like 2 epochs